server/bin/pbench-dispatch.sh

```diff
@@ -239,81 +241,53 @@ while read tbmd5; do
-        # We have a bearer token for the PUT API of a "new" Pbench Server,
-        # invoke the `pbench-results-push` agent CLI to send it to the
-        # server (configuration of that server handled elsewhere).
-
-        # All tar balls pushed to the "new" Pbench Server are made public just
-        # like they are all public on the current "old" Pbench Server.
-        push_options="--token ${put_token} --access=public"
-        push_options+=" --metadata=global.server.legacy.version:${server_version}"
-        push_options+=" --metadata=global.server.legacy.sha1:${server_sha1}"
-        push_options+=" --metadata=global.server.legacy.hostname:${server_hostname}"
-        satellite=${controller%%::*}
-        if [[ "${controller}" != "${satellite}" ]]; then
-            push_options+=" --metadata=server.origin:${satellite}"
-        fi
-        pbench-results-push ${tb} ${push_options}
-        sts=${?}
-        if [[ ${sts} -ne 0 ]]; then
-            log_warn "${TS}: 'pbench-results-push ${tb} ${push_options}' failed, code ${sts}"
-        fi
+    # We have a bearer token for the PUT API of a "new" Pbench Server, invoke
+    # the `pbench-results-push` agent CLI to send it to the server
+    # (configuration of that server handled elsewhere).
+
+    # All tar balls pushed to the "new" Pbench Server are made public just
+    # like they are all public on the current "old" Pbench Server.
+    push_options="--token ${put_token} --access=public"
+    push_options+=" --metadata=global.server.legacy.version:${server_version}"
+    push_options+=" --metadata=global.server.legacy.sha1:${server_sha1}"
+    push_options+=" --metadata=global.server.legacy.hostname:${server_hostname}"
+    satellite=${controller%%::*}
+    if [[ "${controller}" != "${satellite}" ]]; then
+        push_options+=" --metadata=server.origin:${satellite}"
     fi
-
-    # Make sure that all the relevant state directories exist
-    mk_dirs ${controller}
+    pbench-results-push ${tb} ${push_options}
     sts=${?}
     if [[ ${sts} -ne 0 ]]; then
-        log_error "${TS}: Creation of ${controller} processing directories failed for ${tb}: code ${sts}"
+        log_error "${TS}: 'pbench-results-push ${tb} ${push_options}' failed, code ${sts}"
         (( nerrs++ ))
         continue
     fi
 
-    # First, copy the small .md5 file to the destination. That way, if
-    # that operation fails it will fail quickly since the file is small.
-    cp -a ${tbmd5} ${dest}/
+    # Ensure the SELinux context is properly set.
+    tbname=${resultname}.tar.xz
+    restorecon ${dest}/${tbname} ${dest}/${tbname}.md5
     sts=${?}
     if [[ ${sts} -ne 0 ]]; then
-        log_error "${TS}: Error: \"cp -a ${tbmd5} ${dest}/\", status ${sts}"
+        # This likely means the PUT API failed unexpectedly.
+        log_error "${TS}: Error: \"restorecon ${dest}/${tbname} ${dest}/${tbname}.md5\", status ${sts}"
         (( nerrs++ ))
         continue
     fi
```

Comment:
The `pbench-results-push` operation at line 258 is what creates the `${dest}/${tbname}` file, right?  This leads to two questions:
- Why is the `restorecon` command (at line 268) being done by _this_ script doing instead of by the code which actually creates the files?
- Is the file creation operation really synchronous?  That is, can we really depend on the `${dest}/${tbname}` existing by the time we reach line 268?  This assumption seems fragile and/or limiting, so it would be good if we could work around this (e.g., if we move the `restorecon` operation to the Server, does that mean we don't have to worry about synchronicity in this module?).

Reply:
> Why is the `restorecon` command ...

Because the `main` like code doesn't care about SELinux context and can't, it is run inside a container.

> Is the file creation operation really synchronous?

Yes, it had better be.  The whole idea of the PUT API is that a client is "putting" their valuable data into the pbench server and that data had better be securely stored before the API returns.

Reply:
Well, "synchronous" is a big word here. I think that depends on the container and filesystem behaviors. Our upload `PUT` is a fully synchronous API; and if, in the future, we decide to make it asynchronous, we'd need a wait mechanism in `pbench-results-push` to know when it was finished. For now, that doesn't make any sense. But does the fact that the files have been created in the "technically shared" `/srv/pbench` filesystem inside the 1.0 container mean they're already visible outside the container, or even inside a "next-door" 0.69 passthrough container? Well, yeah; maybe that's an issue?

As for SE linux: the `PUT` code (actually cache manager) *does* do the `restorecon`, however we run the 1.0 Server container without selinux, so it's disabled. And there's potentially another issue with managing the selinux context across containers?

Reply:
> the PUT code (actually cache manager) does do the restorecon, however we run the 1.0 Server container without selinux, so it's disabled

Ah, yes.  Hmmm....

> > Is the file creation operation really synchronous?
>
> Yes, it had better be. The whole idea of the PUT API is that a client is "putting" their valuable data into the pbench server and that data had better be securely stored before the API returns.

There's no question about the security of the storage...the question is about how and when the results of that storage become visible to the client.

The client is invoking an HTTP `PUT` method -- if the client were using another RESTful API call to access the file, then we would have much more latitude -- however, this script is instead using the file system to access the resulting file directly.  That's an asymmetry which borders on a violation of modularity...and it's certainly predicated on a number of assumptions about how the API is implemented.

Perhaps that's OK.  It's certainly expedient, and hopefully it won't be required for the long term.  But, we need to be cognizant of what the assumptions are in this code and what the derived requirements are which they impose on the Pbench Server code going forward.

Specifically, it must be the case that the file is in the expected place before the `PUT` method returns (e.g., we cannot queue it up to an S3 service or otherwise defer the disposition)...which is something that we should probably be noting in the code in the `main` branch.